vllm/spec_decode/multi_step_worker.py

```diff
@@ -11,6 +12,68 @@
 from vllm.worker.worker import Worker
 
 
+class NGramDrafter():
+
+    def __init__(self, device, vocab_size):
+        self.ngram_len = 2
+        self.device = device
+        self.vocab_size = vocab_size
+
+    def spec_infer(
+        self,
+        seq_group_metadata_list: List[SequenceGroupMetadata],
+        num_steps: int,
+    ) -> Optional[List[SamplerOutput]]:
+
+        arr = []
+        has_spec_out = False
+        for seq_group_metadata in seq_group_metadata_list:
+            seq_data = next(iter(seq_group_metadata.seq_data.values()))
+
+            input_ids = torch.from_numpy(
+                np.asarray(seq_data.get_token_ids(), np.int64))
+            ngram_tensor = input_ids[-1 * self.ngram_len:]
```

Comment:
In this implementation we only match n-gram equals to the given size. However, we could reduce n-gram size if no match so that we could have more speculative proposals.

Reply:
add the logic for checking window between ngram_prompt_lookup_min and ngram_prompt_lookup_max